File: examples/03-interactions/go/signals_queries.go
package interactions

import (
	"go.temporal.io/sdk/workflow"
)

// Signals arrive on a channel you read with workflow.GetSignalChannel; queries
// are pull handlers you register with workflow.SetQueryHandler. The Workflow
// blocks in workflow.Await (the Go equivalent of Workflow.await) until a Signal
// flips the state.
func HumanApprovalWorkflow(ctx workflow.Context, requestID string) (string, error) {
	state := "WAITING"

	if err := workflow.SetQueryHandler(ctx, "currentState", func() (string, error) {
		return state, nil
	}); err != nil {
		return "", err
	}

	approvals := workflow.GetSignalChannel(ctx, "approve")
	workflow.Go(ctx, func(ctx workflow.Context) {
		var approver string
		approvals.Receive(ctx, &approver)
		state = "APPROVED by " + approver
	})

	_ = workflow.Await(ctx, func() bool {
		return len(state) >= 8 && state[:8] == "APPROVED"
	})
	return state, nil
}

File: examples/03-interactions/python/signals_queries.py
from temporalio import workflow


# Signal/query handlers are methods on the @workflow.defn class, the Python
# equivalent of @SignalMethod / @QueryMethod. The @workflow.run method blocks in
# workflow.wait_condition until a Signal flips the state (like Workflow.await).
@workflow.defn
class HumanApprovalWorkflow:
    def __init__(self) -> None:
        self._state = "WAITING"

    @workflow.run
    async def run(self, request_id: str) -> str:
        await workflow.wait_condition(lambda: self._state.startswith("APPROVED"))
        return self._state

    @workflow.signal
    def approve(self, approver: str) -> None:
        self._state = "APPROVED by " + approver

    @workflow.query
    def current_state(self) -> str:
        return self._state
